frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx

```diff
@@ -1,47 +1,92 @@
+      return HdsNumberInput;
+
+    case 'textArea':
+      return HdsTextArea;
+
+    case 'text':
+    default:
+      return HdsTextInput;
+  }
+}
+
+export type TextInputProps = {
+  validation?: RegisterOptions<Application>;
+  id: NonNullable<Parameters<UseFormRegister<Application>>[0]>;
+  type?: $TextInputProps['$type'],
+  helperFormat?: string,
+} & GridCellProps;
+
+const TextInput : React.FC<TextInputProps> = ({
   id,
-  validation,
-  ...rest
-}: InputProps & TextInputProps): ReturnType<typeof HdsTextInput> => {
+  validation = {},
+  type = 'text',
+  helperFormat,
+  ...$gridCellProps
+}) => {
+  const { t } = useTranslation();
   const {
-    translateLabel,
-    translateError,
     register,
-    formState: { errors },
-  } = useApplicationForm();
-  const { application, isLoading } = useApplicationApi();
+  } = useFormContext<Application>();
+
+  const { getValue, getError, fieldName } = useApplicationFormField<string>(id);
+  const errorType = getError()?.type;
+
+  const errorText = React.useMemo((): string | undefined => {
+    if (!errorType) {
+      return undefined;
+    }
+    const error = t(`common:application.form.errors.${errorType}`);
+    const helperText = helperFormat ? `${t('common:application.form.helpers.format')}: ${helperFormat}` : undefined;
+    if (['pattern','required'].includes(errorType) && helperText) {
+      return `${error}. ${helperText}`;
+    }
+    return error;
+  },[t,errorType, helperFormat]);
+
+  // TODO: This can be removed after backend supports invalid values in draft save
+  const setValueForBackend = React.useCallback((newValue: string) =>
+    // getError does not always update: https://github.com/react-hook-form/react-hook-form/issues/2893
+    // if value hasnt changed (getValue is same as new value), then error is present and invalid value is changed to undefined
+    // to prevent backend to fail
+    getError() && getValue() === newValue ? undefined : newValue,
+   [getError,getValue]);
 
-  const defaultValue = application?.[id] ? String(application[id]) : '';
-  const hasError = Boolean((errors?.[id] as FieldError)?.type);
 
   return (
-    <$TextInput
-      {...rest}
-      {...register(id, validation)}
-      id={id}
-      data-testid={id}
-      name={id}
-      disabled={isLoading}
-      required={Boolean(validation.required)}
-      max={validation.maxLength ? String(validation.maxLength) : undefined}
-      defaultValue={defaultValue}
-      errorText={hasError ? translateError(id) : undefined}
-      label={translateLabel(id)}
-    />
+    <$GridCell {...$gridCellProps}>
+      <$TextInput
+        as={getComponentType(type)}
+        {...register(id, {...validation, setValueAs: setValueForBackend})}
+        $type={type}
+        key={id}
+        id={id}
+        data-testid={id}
+        name={id}
+        required={Boolean(validation.required)}
+        max={validation.maxLength ? String(validation.maxLength) : undefined}
+        defaultValue={getValue()}
```

Comment:
Is this also an uncontrolled input on purpose?

Reply:
Yes, react-hook-form embraces uncontrolled inputs... This does not work if using value instead of defaultValue.